azure/ConsiderationsForServiceDesign.md

```diff
@@ -229,6 +229,33 @@ An important distinction between RELO and status monitor LROs is that there is a
 status monitor LRO, whereas the status of all RELO operations is combined into the status of the resource.
 So status monitor LROs are "one-to-one" with their operation status, whereas RELO-style LROs are "many-to-one".
 
+## Errors
+One of the most important parts of service design is also one of the most overlooked.  The errors returned by your service are a critical part of your developer experience.  Your service and your customer's application together form a distributed system.  Errors are inevitable, but well-designed errors can help you avoid costly customer support incidents by empowering customers to self-diagnose problems.
+
+First, you should always try to design errors out of existence if possible.  You'll get a lot of this for free by following the Guidelines.  Some examples include:
+- Idempotent APIs solve a whole class of network issues where customers have no idea how to proceed if they send a request to the service but never get a response.
+- Accessing resources from multiple microservices can quickly lead to complex race conditions so conditional requests provide optimistic concurrency for safe usage.
+- Reframing the purpose of an API can obviate some errors.  This is most often specific to your operations, but an example from the Guidelines is thinking about `DELETE`s as _"ensure no resource at this location exists"_ so they can return an easier to use `204` instead of _"delete this exact resource instance"_ which would fail with a non-idempotent `404`.
+
+There are two types of errors returned from your service and customers handle them differently.
+- Usage errors where the customer is calling your API incorrectly.  The customer can easily make these errors go away by fixing their code.  We expect most usage errors to be found during testing.
+- Runtime errors that can't be prevented by the customer and need to be recovered from.  Some runtime errors like `429` throttling will be handled automatically by client libraries, but most will be situations like a `409` conflict requiring knowledge about the customer's application to remedy.
+
+We provide appropriate [HTTP status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) for customers to handle errors generically and error code strings in our common error schema and the `x-ms-error-code` header for customers to handle errors specifically.  As an example, consider what a customer would do when trying to get the properties of a Storage blob:
```

Comment:
What is the recommendation when we have both long-running and "sync" versions of the same operations?  The sync operations may return various 4xx.  But the long-running version would need to return 200 and embed the status code in the inner error.  Thus, should the top level error code more or less reflect HTTP status codes?  Or should we have more specific errors?

Specifically, for [Form Recognizer](https://docs.microsoft.com/en-us/azure/applied-ai-services/form-recognizer/preview-error-guide), should the top-level error codes be InvalidRequest/InvalidArgumen/NotFound/... or ContentSourceNotAccessible/InvalidContentSourceFormat/ModelNotFound/...?

Examples of best practcies will be useful.

Reply:
I don't see the sync/async aspect as any different.  I'd return a `404` with an error code `ModelNotFound` from the sync API and embed both of those values in your LRO's `GetStatus`.

Reply:
When the LRO fails, it is my understanding that GetStatus should return 200 since the GetStatus operation itself succeeded.  Instead, we return the error via the _error_ property.  Thus, there is no place to return 404.  Is my understanding incorrect?

Reply:
I think we should just include the status code as an additional property inside the 200's error property.  We do that in many of our batching APIs for a similar example that's worked well.  If folks agree, I think we'd want that recommendation in the LRO section.

Reply:
Captured this thread as issue #317